src/cart/swim/swim.c

```diff
@@ -821,8 +827,12 @@ swim_progress(struct swim_context *ctx, int64_t timeout)
 		D_GOTO(out_err, rc = 0); /* Ignore this update */
 
 	now = swim_now_ms();
-	if (timeout > 0)
+	if (timeout > 0) {
```

Comment:
I suppose this 'timeout' is the passed in 'timeout' value from crt_progress(), right? It's by default 1 ms when the xstream is idle, is it good for SWIM? Or do we need a different timeout value for swim?

Reply:
yes,  'timeout' is the passed from crt_progress(). It's zero (0) by default to exit as fas as possible. Now I have separate xstream to progress only SWIM. So, it would be good don't eat CPU cycles by waiting next SWIM event. Here I Calculate next wakeup time and sleep to it instead of busy looping.

Reply:
The crt_progress() passed in timeout value is 1 ms by default for an idle xstream. I don't quite follow here, you calculate next wakeup time when 'time > 0' or when 'time == 0'?

Reply:
@liw what you said above looks better than below usleep() method. 
usleep() below can do nothing, but with a reasonable timeout passed back to crt_progress() -> crt_hg_progress() -> HG_Progress() can wait on HG internal wait obj/fd that will not miss possible network activity.

Reply:
There are two different "timeout" arguments here. One is passed into crt_progress(), the second is passed in swim_progress().   The second timeout originally was zero (0) to avoid any loops and sleeps inside the progress if possible. Now I check if SWIM have a separate stream it can loop and sleep inside progress callback. So, those changes accommodate this possibility. Please look at [ctr_swim.c](https://github.com/daos-stack/daos/pull/6635/files/475f30d326bf8885f065e5724096429aa0419040#diff-3692220b800418eb141a001eb2f22702fa1e603c9c9883beb49d46a73d38e4a0) in 698 line for details.

Reply:
I agree this is good idea to pass the knowledge about timeout to crt_progress() back. Will try to realize this...